pygmt/clib/session.py

```diff
@@ -793,11 +793,68 @@ def put_vector(self, dataset, column, vector):
                 )
             )
 
+    def put_strings(self, dataset, family, strings):
+        """
+        Attach a numpy 1D array of dtype str as a column on a GMT dataset.
+
+        Use this function to attach string type numpy array data to a GMT
+        dataset and pass it to GMT modules. Wraps ``GMT_Put_Strings``.
+
+        The dataset must be created by :meth:`~gmt.clib.Session.create_data`
+        first.
+
+        .. warning::
+            The numpy array must be C contiguous in memory. If it comes from a
+            column slice of a 2d array, for example, you will have to make a
+            copy. Use :func:`numpy.ascontiguousarray` to make sure your vector
+            is contiguous (it won't copy if it already is).
+
+        Parameters
+        ----------
+        dataset : :class:`ctypes.c_void_p`
+            The ctypes void pointer to a ``GMT_Dataset``. Create it with
+            :meth:`~gmt.clib.Session.create_data`.
+        family : str
+            The family type of the dataset. Can be either ``GMT_IS_VECTOR`` or
+            ``GMT_IS_MATRIX``.
+        strings : numpy 1d-array
+            The array that will be attached to the dataset. Must be a 1d C
+            contiguous array.
+
+        Raises
+        ------
+        GMTCLibError
+            If given invalid input or ``GMT_Put_Strings`` exits with status !=
+            0.
+
+        """
+        c_put_strings = self.get_libgmt_func(
+            "GMT_Put_Strings",
+            argtypes=[
+                ctp.c_void_p,
+                ctp.c_uint,
+                ctp.c_void_p,
+                ctp.POINTER(ctp.c_char_p),
+            ],
+            restype=ctp.c_int,
+        )
+
+        strings_pointer = (ctp.c_char_p * len(strings))()
+        strings_pointer[:] = np.char.encode(strings)
+
+        status = c_put_strings(
+            self.session_pointer, self[family], dataset, strings_pointer
+        )
```

Comment:
With the upstream PR https://github.com/GenericMappingTools/gmt/pull/3718 and some changes in the Python code (I don't push the changes yet), the `GMT_Put_Strings` wrapper works well. Please see the full discussion in that PR, especially comments https://github.com/GenericMappingTools/gmt/pull/3718#issuecomment-664084529 and https://github.com/GenericMappingTools/gmt/pull/3718#issuecomment-664089340.

For GMT_Put_Vector, GMT_Put_Matrix, GMT_Put_Strings, they don't duplicate the vector, matrix or strings by default. So if the vector, matrix, or strings are freed in Python, we should see some unexpected behaviors or crashes. The situation never happens for GMT_Put_Vector and GMT_Put_Matrix, but happens for GMT_Put_Strings.

vectors/matrix and strings are passed using different ways. Here is how we pass vectors,
```
vector_pointer = vector.ctypes.data_as(ctp.c_void_p)
```
here is how we pass strings)
```
        strings_pointer = (ctp.c_char_p * len(strings))()
        strings_pointer[:] = np.char.encode(strings)
```
As I understand it, `vector_pointer` is a pointer pointing to the original `vector` memory, but `strings_pointer` looks like a new pointer (I don't understand the codes. Maybe I'm wrong here) and is freed when exiting the function.

For strings, if we have a way to pass data just like vectors or matrix, then we probably don't need PR #3718, which duplicates the strings.

Reply:
Ideally we could just do `strings_pointer = strings.ctypes.data_as(ctp.c_char_p)`. The only difference seems to be `ctp.c_void_p` and `ctp.c_char_p`. No idea about how memory is freed/handled by Python/NumPy, I would think that they're both just a *direct* pointer to the same thing in memory (Printing the `c_char_p`/`c_void_p` pointer yields the same number (something like `123456789`) from what I recall), but I'm not too sure here either. Might have to wait for [Hurrican Douglas](https://en.wikipedia.org/wiki/Hurricane_Douglas_(2020)) to pass over Hawaii before we can test this properly.